### ISSUE TITLE
Problems with hi-res / retina displays

### ISSUE THREAD

**Issue by sjackso** [angel2d/angel2d#18]
I am using Angel2d on a retina macbook pro.  Although Angel2d has some support for high resolution screens (an excellent feature in a cross-platform game engine!), there are some issues.
- GWEN UI elements are drawn with incorrect sizesand, sometimes, at incorrect locations.
- The mouse coordinates for picking the GWEN UI elements are significantly shifted from where they should be.
- If a window is moved from a retina screen to a conventional screen, or vice versa, GWEN UI elements disappear entirely.
- World::SetSideBlockers works incorrectly -- blockers exist but are far outside the window boundaries.

I will follow up with some thoughts on fixing these issues.

Screenshots from IntroGame:

![gwenerr1](https://f.cloud.github.com/assets/2493183/2380554/29d47998-a8ad-11e3-83fe-0f1b3be54800.png)

![gwenerr2](https://f.cloud.github.com/assets/2493183/2380561/57b48db2-a8ad-11e3-97b2-f7eedaba51fb.png)

**Comment by sjml** [comment]
Thanks for looking into this -- I've always had a hunch that there were gaps in Retina support, but lacking a Retina MBP I haven't been able to test. Some of these could probably be solved speculatively, but any direct patches you might have would be appreciated.

**Comment by sjackso** [comment]
One underlying problem is that high res screens make a liar out of MathUtil::ScreenToWorld().  The first lines of this function are:

``` c++
Vector2 MathUtil::ScreenToWorld(int x, int y)
{
    if (theWorld.IsHighResScreen())
    {
        x *= 2;
        y *= 2;
    }
    ...
```

i.e., on a retina screen the ScreenToWorld method always behaves as if you asked for the world coordinates of the (2x, 2y)th pixel.  As far as I can tell, this is only correct for code that knew the _requested_ size of the window, but hasn't noticed that the _actual_ window size is twice what was requested.  For code that looks up x and y based on the actual screen dimensions (such as SetSideBlockers()), this multiplication produces incorrect results.

The labels in the IntroGame are an interesting example of the issues.  The file label, a TextActor, places itself with MathUtil::ScreenToWorld(), so it doesn't know whether the window is high resolution.  The info string, drawn manually with DrawGameText(), checks the screen size directly.   If I comment out the if block above, so that ScreenToWorld never scales its arguments, then the screen looks like this:

![fix](https://f.cloud.github.com/assets/2493183/2380689/25e3a928-a8b0-11e3-86c0-8d12298ff15e.png)

So a first question for fixing the issue is: where should the scaling take place?  My intuition is that ScreenToWorld() should never scale its arguments-- rather, calling code should be paying attention to the actual size of the window, and not assuming that the size given to World::Initialize() is necessarily correct.  But this will break any actors that are manually aligned using MathUtil::ScreenToWorld, such as the file label in the image above.

Note that removing the scaling from ScreenToWorld does not fix the GWEN issues, so those will have to be addressed separately.

**Comment by sjml** [comment]
The trick is that you usually _do_ want to treat the coordinates as "logical" rather than literal pixels -- this follows the example Apple sets on iOS (where the hi-res code originated) and in what I can see of retina desktop code. 

Angel should be reporting the window size as logical pixels instead of the actual width of the viewport, so at the end of the day the user doesn't care whether they're running on a retina screen or not. Is that not the case?

**Comment by sjml** [comment]
In answer to your question: ScreenToWorld definitely _should_ be scaling its inputs, since that allows the calling code to not care if its on a retina display or not. Of course, not all of Angel's calling code is smart enough to realize this, and that will need to be addressed.

**Comment by sjackso** [comment]
I am not sure about what is usual best practice in iOS and desktop retina code, but you're probably right.  

If we take that approach, then Angel would speak entirely in terms of logical pixels, and would do its best to hide the physical pixels from client code (though code that needs physical pixels, e.g. calls into opengl, will have to check whether the screen is high res).

In this case, then the Camera should be reporting logical instead of physical pixels in methods like GetWindowHeight().

**Comment by sjml** [comment]
Yep, the GetWindowHeight was likely just an oversight. :-/

**Comment by sjackso** [comment]
Changing GetWindowHeight and GetWindowWidth to return logical pixels does not immediately have the effect we want:

![enormity](https://f.cloud.github.com/assets/2493183/2380899/c8927894-a8b4-11e3-872d-67d9d187a2e8.png)

... will investigate further

**Comment by sjml** [comment]
Yeah, as you've seen, it's all pretty intertwined in annoying ways. Oh for the simple days when a pixel was a pixel. 

This is one of those annoying issues where, if I had the appropriate hardware, I could probably resolve it engine-wide in a couple of hours. But without a Retina display, it's all kind of guesswork, at least the way Angel3 has all this implemented. :-/

**Comment by sjml** [comment]
That said, please continue your investigations! Look forward to seeing if you can nail down a set of changes that solves it.

**Comment by iluuu1994** [comment]
@sjml I noticed the same thing. I can help :)
I'm in ❤️ with this engine, bugger that it doesn't get more attention.

**Comment by sjml** [comment]
It gets as much attention as I can give. :) Full time job + other side projects, sadly, draw a lot of my time. And seriously, if I had a retina machine, this would probably be fixed in a couple hours. :( Patches welcome!

**Comment by iluuu1994** [comment]
@sjml I was talking about a bigger community ;)
I will try out some things when I find the time.

Ps, I think there are [ways](http://stackoverflow.com/questions/12124576/how-to-simulate-a-retina-display-hidpi-mode-in-mac-os-x-10-8-mountain-lion-on) of simulating a retina display.

**Comment by ErlerPhilipp** [comment]
It's almost the same with DPI-scaling in Windows. The easiest solution there is to set the "Enable DPI Awareness" property in the ClientGame project settings -> Configuration Properties -> Manifest Tool -> Input and Output to Yes (or whatever your version of Visual Studio offers).
